test/distributed/test_c10d.py

```diff
@@ -3499,6 +3510,69 @@ def test_sparse_gradients(self):
     def test_sparse_gradients_grad_is_view(self):
         self._test_sparse_gradients(gradient_as_bucket_view=True)
 
+    def _run_and_verify_mixed_dense_sparse_gradients(self, vanilla_model, ddp_model):
+        mult = 2
+        batch_size = mult * self.world_size
+        criterion = nn.CrossEntropyLoss()
+        input = torch.randint(0, 10, [batch_size, 10])
+        target = torch.randint(0, 10, [batch_size])
+
+        criterion(vanilla_model(input), target).backward()
+
+        partial_input = input.split(mult)[self.rank]
+        partial_target = target.split(mult)[self.rank]
+
+        loss = ddp_model(partial_input)
+        # communication hook does not work on gradients adding hook to loss
+        loss.register_hook(lambda x: x / self.world_size)
+
+        criterion(loss, partial_target).backward()
+
+        for vanilla_parameter, ddp_parameter in zip(vanilla_model.parameters(), ddp_model.parameters()):
+            self.assertEqual(vanilla_parameter.grad, ddp_parameter.grad)
+
+    def _test_mixed_dense_sparse_gradients(self, gradient_as_bucket_view=False):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
+        torch.manual_seed(1337)
+
+        vanilla_model = MixedDenseSparseGradientModule()
+        ddp_model = DistributedDataParallel(
+            copy.deepcopy(vanilla_model),
+            process_group=process_group,
+            gradient_as_bucket_view=gradient_as_bucket_view,
+        )
+
+        # "get_future" API does not support gloo backend, see GH Issue #42048.
+        # Instead, we wait for an allreduce work, and write its result to a Future.
+        def allreduce_hook_gloo(
```

Comment:
curious, why do we need a hook here? Does DDP's default comm hit any error?

Reply:
resolved from based on the dense-sparse conversation.